### ISSUE TITLE
can't use tab-completion with Yoink Mac app

### ISSUE THREAD

**Issue by goodloop** [robbyrussell/oh-my-zsh#5292]
encounter the same problem as [this](http://apple.stackexchange.com/questions/179282/tab-completion-in-terminal-with-alias)

attached omz-debug log.
[gist file](https://gist.github.com/goodloop/73fcc2f212b0766de8f1daa45a1c1ad0).
please help.

----
Transcription from the above link:

I'm running a zsh terminal under Yosemite and have defined a simple alias to place files in Yoink:

```
$ alias yoink='open -a Yoink'
$ yoink my-file-to-yoink.txt
```

This works great but for some unknown reason the tab completion to specify the file does not work after yoink. It simple says `-- no matches found --`.

Is there a way I can use tab completing in combination with this alias?

Thanks.

ps: Not sure if it's related, but I'm using zprezto.

**Comment by mcornella** [comment]
This shouldn't happen because zsh's `open` command completion works by default. What happens if you try that without the alias, _i.e._ doing `open -a Yoink [TAB]`?

**Comment by goodloop** [comment]
I tried "open -a Yoink", doesn't work either.
I also tried to restore original .zshrc  to disable oh-my-zsh, and  It works. So I try to get some hint here

**Comment by mcornella** [comment]
> I also tried to restore original .zshrc to disable oh-my-zsh, and It works. So I try to get some hint here

Ok that changes the situation. Please try running `setopt aliases` with the zshrc file with oh-my-zsh enabled?

**Comment by goodloop** [comment]
no luck with `setopt aliases`.

but seems `setopt completealiases` works. Does it make sense?

**Comment by mcornella** [comment]
It makes sense that `completealiases` works, but it will break the rest of aliases completions, because it has the effect of treating an alias as a command. Therefore `yoink [TAB]` will search for a `_yoink` completion function and when it doesn't find one it will just offer to complete files (that's the default).

There must be some other problem with the `_open` completion function. If you try `open [TAB]` is there any output? After that, what does `which _open` output?

**Comment by goodloop** [comment]
I restore to `setopt aliases` first.

the output of `which _open`

```
_open () {
	local curcontext="$curcontext" state line expl
	_arguments -C '-a[specify application]: :->open_mac_applications' '-e[open with TextEdit]' '-f[reads input from standard input and opens with TextEdit]' '*: :->open_files'
	case "$state" in
		(open_mac_applications) _alternative "commands: :_mac_applications" "files:: _open_absolute_application_path" ;;
		(open_files) local app
			if [[ -n "$words[(r)-a]" ]]
			then
				app="${(Q)words[words[(i)-a] + 1]}"
			elif [[ -n "$words[(r)-e]" || -n "$words[(r)-f]" ]]
			then
				app="Text Edit"
			fi
			if [[ -n "$app" ]]
			then
				_wanted files expl "file for $app" _mac_files_for_application "$app"
			else
				_webbrowser
			fi ;;
	esac
}
```

**Comment by mcornella** [comment]
That seems to be ok... Try with `rm ~/.zcompdump*` and restarting the shell.

**Comment by goodloop** [comment]
no luck:(

**Comment by mcornella** [comment]
That is really weird... Try running `functions -t _open` (turn on tracing for open completion function), then `yoink [TAB]` and paste the output here.

**Comment by goodloop** [comment]
[yoink TAB log](https://gist.github.com/goodloop/5b42861cfc601dd5872ba7063daa3a8b)

**Comment by apjanke** [comment]
## TL;DR

I can reproduce. I don't think this is OMZ. The `_open` function (provided by `zsh` itself) is working normally, but Yoink.app's metadata doesn't declare it handles any input file types, so `_open` doesn't think there are any valid file candidates for completion. Try reporting this to the Yoink developers, or write a custom `_yoink` function.

##  Details

I can reproduce this. Without Yoink installed, completing `open -a Yoink ...` will list a bunch of files as possible completions. But with Yoink 3.2 installed, completing `open -a Yoink ...` offers zero files. I also get the same behavior if I turn off Oh My Zsh, but still run `compinit` as part of my `zsh` startup.

This may be a file type and application metadata thing. When doing `open -a Foo ...`, `_open` will only list files whose extensions are valid for the particular application you're completing against. This is done by calling the `_mac_files_for_application` function.

```
if [[ -n "$app" ]]
            then
                _wanted files expl "file for $app" _mac_files_for_application "$app"
            else
```

Here's that function's logic. It looks into the installed application's `Info.plist` to see what file types it declares that it handles.

```
$ which _mac_files_for_application
_mac_files_for_application () {
	local -a opts
	zparseopts -D -a opts q n 1 2 P: S: r: R: W: X+: M+: F: J+: V+:
	local app_path
	_retrieve_mac_apps
	app_path="${_mac_apps[(r)*/$1(|.app)]:-$1}"
	local -a glob_patterns
	glob_patterns=()
	if [[ -f "$app_path/Contents/Info.plist" ]]
	then
		local -a exts types
		_mac_parse_info_plist
		if [[ -n "$exts[(r)\*]" ]]
		then
			glob_patterns=("*")
		else
			if (( #exts != 0 ))
			then
				glob_patterns+=("*.(${(j/|/)exts})(N)")
			fi
			if (( #types != 0 ))
			then
				glob_patterns+=("^*.[[:alnum:]]##(.Ne:_mac_rsrc_check:)")
			fi
		fi
	else
		glob_patterns=("*")
	fi
	case ${#glob_patterns} in
		(0) return 1 ;;
		(1) _files "$opts[@]" -g "$glob_patterns[1]" ;;
		(*) _files "$opts[@]" -g "{${(j/,/)glob_patterns}}" ;;
	esac
}
[~/.oh-my-zsh on ⇄ master]
$
```

And here's the logic for parsing that info plist.

```
$ which _mac_parse_info_plist
_mac_parse_info_plist () {
	local s='
    BEGIN { RS="<" }
      /^key>/ { sub(/key>/, ""); reading_key=$0 }
      /^string>/ {
        sub(/string>/, "")
        if (reading_key == "CFBundleTypeExtensions") exts=exts " \"" $0 "\""
        if (reading_key == "CFBundleTypeOSTypes") types=types " \"" $0 "\""
      }
      END {
        print "exts=(" exts ")\ntypes=(" types ")"
      }
    '
	command awk $s "$app_path/Contents/Info.plist" | while read
	do
		eval "$REPLY"
	done
}
[~]
$
```

I took a look in Yoink.app's `Contents/Info.plist` file. It does not contain either `CFBundleTypeExtensions` or `CFBundleTypeOSTypes` elements. Basically, I think that means it's not declaring any types of files that it handles as inputs. So `_open` doesn't think there any files that are valid for completion candidates.

I think this means that `_open` is acting as designed, and the Yoink app's metadata is interacting with it in an inconvenient way. If Yoink.app really was designed to take any files as input from the command line or `open` invocation, it should probably declare them in its `Info.plist`. You may want to contact the Yoink developers about this to see if they can adjust the `Info.plist`, if this is a supported behavior for Yoink.

You could also fix this yourself be writing a custom `_yoink` function and getting that on your `$FPATH` for Zsh's completion to use. To get this to work, you will need to replace your `yoink` alias with a `yoink()` function that does the same things: completion "looks inside" aliases, but not inside functions, and like Marc said, you don't want to turn on `setopt complete_aliases`.

> ps: Not sure if it's related, but I'm using zprezto.

Prezto and OMZ are alternative frameworks to do the same shell configuration work. Running both of them could have unforeseen interactions and breakage, and is not supported. But I don't think that's relevant here.

**Comment by mcornella** [comment]
Thanks for the summary Andrew. Your post should be enough to report a bug. Looking at [Yoink's website](http://eternalstorms.at/yoink/) I don't exactly get what `open -a Yoink <file>` would do since it's a drag & drop app. Does that do something at least?

**Comment by apjanke** [comment]
Yes. `open -a Yoink <file>` adds it to the Yoink staging area, just like it had been drag-and-dropped there. Given that behavior, I would think they want to declare `'*'` as a handled extension type in `CFBundleTypeExtensions`, or something similar.

**Comment by goodloop** [comment]
Thanks, all guys.  it will put file into dragged area, and waiting  for drop to the destination positions

**Comment by mcornella** [comment]
Ok. I'm gonna contact [the developer](https://twitter.com/eternalstorms) via Twitter and let him know. In the meantime, a simple `compdef _files yoink` should do the job.

**Comment by goodloop** [comment]
setopt aliases is not needed any more,right?

**Comment by mcornella** [comment]
He has a Github account as well. @eternalstorms please have a look at [this bug report](https://github.com/robbyrussell/oh-my-zsh/issues/5292#issuecomment-239318879) for Yoink's zsh completion. It only requires modifying the `Info.plist` file.

**Comment by mcornella** [comment]
> setopt aliases is not needed any more,right?

Nope. It was a guess but it wasn't that the problem so you can get rid of it.

**Comment by eternalstorms** [comment]
Hi.
Thank you for the heads-up on this, I wasn't aware of this issue at all.
I'll fix the Info.plist with the next update, it should be out some time next week.

Thanks again!
- Matt

**Comment by eternalstorms** [comment]
The update is available on the Mac App Store now - if you have some time, please let me know if it's working correctly now. Thank you.

**Comment by apjanke** [comment]
Is the demo on the website updated? That's what I used to reproduce; I haven't actually purchased Yoink.

**Comment by goodloop** [comment]
@eternalstorms I have updated Yoink, and it works perfect. Thanks very much @mcornella @eternalstorms @apjanke .

**Comment by eternalstorms** [comment]
@apjanke yes, the demo is updated as well.
@goodloop thank you for confirming, I'm glad it's working now :)
